### PR TITLE
Issue #178: Fix malformed HQL when @additional_filters@ is preceded by and

### PR DESCRIPTION
**Pull request by RomanMagnoli**
## Summary

Backport to `release/25.4` (v25) of the fix for ETP-3623.

Fixes a `QuerySyntaxException` thrown when creating records via the EtendoRX Datasource API when the selector HQL contains `and @additional_filters@` and the filter expression evaluates to a non-empty string.

The `else` branch in `buildHQLQuery()` was only replacing the bare `@additional_filters@` placeholder, leaving the preceding `and` keyword in place and producing malformed HQL like `... and AND c.id = '102' ...`.

The fix applies the same two-step replacement already used in the empty-filter branch:

```java
result = hqlQuery.replace("and @additional_filters@", additionalFilters)
    .replace("@additional_filters@", additionalFilters);
```

## Changes

- `SelectorHandlerUtil.java` — fix two-step replacement in `buildHQLQuery()` else branch
- `SelectorHandlerUtilTest.java` — add 3 regression tests covering empty filters, `and`-preceded placeholder, and standalone placeholder

## Jira

ETP-3623

### CODE REVIEW

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 1]
```diff
@@ -868,4 +879,57 @@ public void testSavePrefixFields_MultipleOutfields() throws ReflectiveOperationE
...
+        when(mockTab.getADFieldList()).thenReturn(Collections.singletonList(mockField));
+
+        JSONObject dataInpFormat = new JSONObject();
+        dataInpFormat.put("inpmProductId", "test");
+
+        String result = callBuildHQLQuery(mockSelector, mockTab, mockColumn, "inpmProductId",
+                dataInpFormat, new HashMap<>(), mockRequest);
+
+        assertFalse("HQL must not contain 'and AND'", result.contains("and AND"));
+        assertTrue("HQL must contain the filter clause", result.contains("pl.salesPriceList = true"));
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **Suggestion**

Consider making the check for duplicate 'AND' keywords case-insensitive.

```java
assertFalse("HQL must not contain duplicate AND regardless of case", result.toLowerCase().contains("and and"));
```

**Rationale**: HQL keywords like 'AND' are case-insensitive. Checking for 'and AND' only might miss other malformed variants like 'AND AND' or 'and and'. Testing in a case-insensitive manner is more robust.

File: src/com/etendoerp/etendorx/utils/SelectorHandlerUtil.java [hunk 2]
```diff
@@ -461,7 +461,8 @@ private static String buildHQLQuery(Selector selectorDefined, Tab tab, Column co
         if (StringUtils.isEmpty(additionalFilters.trim())) {
             result = hqlQuery.replace("and @additional_filters@", "").replace("@additional_filters@", "");
         } else {
-            result = hqlQuery.replace("@additional_filters@", additionalFilters);
+            result = hqlQuery.replace("and @additional_filters@", additionalFilters)
+                .replace("@additional_filters@", additionalFilters);
```
Comment (etendobot) [inline src/com/etendoerp/etendorx/utils/SelectorHandlerUtil.java:465]: **Suggestion**

Ensure that 'additionalFilters' consistently includes the leading 'and' connector when it is non-empty, as the replacement of 'and @additional_filters@' will consume the connector from the HQL template.

```java
result = hqlQuery.replace("and @additional_filters@", additionalFilters)
                .replace("@additional_filters@", additionalFilters);
```

**Rationale**: The fix assumes that 'additionalFilters' always starts with the 'and' keyword (e.g., ' and (condition)'). This is a common pattern in Etendo, but if 'additionalFilters' does not start with 'and', this replacement logic would result in missing connectors in the final HQL. Given the PR title, this is the intended behavior to avoid double 'and' keywords.

**Review by etendobot — APPROVED** [review]

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 3]
```diff
@@ -51,6 +51,8 @@ public class SelectorHandlerUtilTest extends WeldBaseTest {
     private static final String FALLBACK_NAME_NORMALIZED = "fallback$name";
     private static final String INP_C_BPARTNER_ID = "inpcBpartnerId";
     private static final String INP_C_BPARTNER_ID_DES = "inpcBpartnerId_DES";
+    private static final String INP_M_PRODUCT_ID = INP_M_PRODUCT_ID;
+    private static final String SALES_PRICE_LIST_FILTER = SALES_PRICE_LIST_FILTER;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Assign the intended string literal value to the constant.

```java
private static final String INP_M_PRODUCT_ID = "inpmProductId";
```

**Rationale**: A static final variable cannot be initialized using its own name. This results in a compilation error ('illegal forward reference' or similar). The constant should be assigned the string literal it is intended to represent, which in this case is 'inpmProductId' based on the refactored lines.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 4]
```diff
@@ -51,6 +51,8 @@ public class SelectorHandlerUtilTest extends WeldBaseTest {
...
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Assign the intended string literal value to the constant.

```java
private static final String SALES_PRICE_LIST_FILTER = "pl.salesPriceList = true";
```

**Rationale**: A static final variable cannot be initialized using its own name. This results in a compilation error. The constant should be assigned the string literal it is intended to represent, which in this case is 'pl.salesPriceList = true' based on the refactored lines.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
+package com.etendoerp.etendorx.utils;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the beginning of the file, before the package declaration.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.etendorx.utils;
```

**Rationale**: New Java source files in Etendo modules must include the standard license header for legal and compliance reasons. This is a mandatory requirement for all com.etendoerp.* modules.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java [hunk 6]
```diff
@@ -0,0 +1,462 @@
+package com.etendoerp.etendorx.utils;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the very top of the new Java file, before the package declaration.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2025 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.etendorx.utils;
```

**Rationale**: Every new Java class in a com.etendoerp.* module must include the standard Etendo license header to comply with legal and organizational standards.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 7]
```diff
@@ -3,172 +3,30 @@
...
 import org.openbravo.model.ad.datamodel.Column;
-import org.openbravo.model.ad.ui.Field;
-import org.openbravo.model.ad.ui.Tab;
-import org.openbravo.userinterface.selector.Selector;
 import org.openbravo.userinterface.selector.SelectorField;
 
 /**
  * Unit tests for SelectorHandlerUtil class.
+ * Covers: processValue, convertRowToJSONObject, fullfillSessionsVariables,
+ * convertToHashMAp, getNormalizedFieldName, getTargetKey.
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **Suggestion**

Verify that the tests for 'buildHQLQuery' and the ETP-3623 scenarios are still executed and documented after the refactoring.

```java
* Covers: processValue, convertRowToJSONObject, fullfillSessionsVariables, buildHQLQuery,
```

**Rationale**: The PR title indicates a fix for HQL generation (Issue #178), but the tests for 'buildHQLQuery' (covering ETP-3623) were removed from this file and are not listed in the Javadoc 'Covers' section. Confirm that these tests were moved to the base class and still provide the necessary coverage.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src/com/etendoerp/etendorx/utils/SelectorHandlerUtil.java (hunk 2 above)
Comment (RomanMagnoli) [reply src/com/etendoerp/etendorx/utils/SelectorHandlerUtil.java:465]: /false-positive

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java (hunk 5 above)
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the beginning of the file, before the package declaration.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.etendorx.utils;
```

**Rationale**: New Java source files in com.etendoerp.* modules must include the standard Etendo license header at the top of the file for legal and compliance reasons.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    String callFullfillSessionsVariables(String whereClause, Map<String, String> db2Input,
+            JSONObject dataInpFormat) throws ReflectiveOperationException {
+        Method method = SelectorHandlerUtil.class.getDeclaredMethod("fullfillSessionsVariables",
+                String.class, Map.class, JSONObject.class);
+        method.setAccessible(true);
+        return (String) method.invoke(null, whereClause, db2Input, dataInpFormat);
+    }
+
+    HashMap<String, String> callConvertToHashMAp(JSONObject dataInpFormat) throws ReflectiveOperationException {
+        Method method = SelectorHandlerUtil.class.getDeclaredMethod("convertToHashMAp", JSONObject.class);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:109]: **Suggestion**

Fix the typo in the method name 'callConvertToHashMAp' to 'callConvertToHashMap'.

```java
    HashMap<String, String> callConvertToHashMap(JSONObject dataInpFormat) throws ReflectiveOperationException {
```

**Rationale**: Typo in the method name ('HashMAp' should be 'HashMap'). Consistent naming improves readability.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java (hunk 6 above)
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the very top of the new Java file, before the package declaration.

```java
/*\n *************************************************************************\n * The contents of this file are subject to the Etendo License\n * (the \"License\"), you may not use this file except in compliance\n * with the License.\n * You may obtain a copy of the License at\n * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt\n * Software distributed under the License is distributed on an\n * \"AS IS\" basis, WITHOUT WARRANTY OF ANY KIND, either express or\n * implied. See the License for the specific language governing rights\n * and limitations under the License.\n * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L\n * All Rights Reserved.\n * Contributor(s): Futit Services S.L.\n *************************************************************************\n */\npackage com.etendoerp.etendorx.utils;
```

**Rationale**: Every new Java class in a com.etendoerp.* module must include the standard Etendo license header to comply with legal and organizational standards. The current diff shows the file starts directly with the package declaration.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 9]
```diff
@@ -3,172 +3,30 @@
...
-    }
-
-    private String callAddFilterClause(Selector selector, HashMap<String, String> hs1,
-            HttpServletRequest request) throws ReflectiveOperationException {
-        Method method = SelectorHandlerUtil.class.getDeclaredMethod("addFilterClause",
-                Selector.class, HashMap.class, HttpServletRequest.class);
-        method.setAccessible(true);
-        return (String) method.invoke(null, selector, hs1, request);
-    }
+public class SelectorHandlerUtilTest extends SelectorHandlerUtilBaseTest {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:30]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Ensure that the base class `SelectorHandlerUtilBaseTest` is included in the PR or that the test continues to extend a valid base class like `WeldBaseTest`.

```java
public class SelectorHandlerUtilTest extends WeldBaseTest { // Or ensure SelectorHandlerUtilBaseTest is included
```

**Rationale**: Extending a non-existent class will cause compilation failures, preventing the project from building. The class `SelectorHandlerUtilBaseTest` was not found in the codebase or the provided diff.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java (hunk 7 above)
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **Suggestion**

Correct the spelling of 'fulfill' and verify if `buildHQLQuery` should still be listed as a covered method, especially since it is relevant to the PR's objective.

```java
 * Covers: processValue, convertRowToJSONObject, fulfillSessionsVariables, buildHQLQuery,
```

**Rationale**: The PR title indicates a fix for malformed HQL, but the updated list of covered methods in the Javadoc excludes `buildHQLQuery`. Additionally, 'fulfill' is misspelled.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 10]
```diff
@@ -3,172 +3,30 @@
...
-import org.openbravo.model.ad.ui.Field;
-import org.openbravo.model.ad.ui.Tab;
-import org.openbravo.userinterface.selector.Selector;
 import org.openbravo.userinterface.selector.SelectorField;
 
 /**
  * Unit tests for SelectorHandlerUtil class.
+ * Covers: processValue, convertRowToJSONObject, fullfillSessionsVariables,
+ * convertToHashMAp, getNormalizedFieldName, getTargetKey.
  * All tests invoke actual production code via reflection to ensure SonarQube coverage.
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:27]: **Suggestion**

Correct the typo 'convertToHashMAp' to 'convertToHashMap' (lowercase 'p' and lowercase 'm' except for the first letter of 'Map').

```java
 * convertToHashMap, getNormalizedFieldName, getTargetKey.
```

**Rationale**: Typographical errors in documentation reduce the professional quality of the code and can mislead developers.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java (hunk 9 above)
Comment (RomanMagnoli) [reply src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:30]: /false-positive

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 11]
```diff
@@ -3,172 +3,30 @@
...
 import org.openbravo.model.ad.datamodel.Column;
-import org.openbravo.model.ad.ui.Field;
-import org.openbravo.model.ad.ui.Tab;
-import org.openbravo.userinterface.selector.Selector;
 import org.openbravo.userinterface.selector.SelectorField;
 
 /**
  * Unit tests for SelectorHandlerUtil class.
+ * Covers: processValue, convertRowToJSONObject, fulfillSessionsVariables,
+ * convertToHashMap, getNormalizedFieldName, getTargetKey.
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:26]: **Suggestion**

Add a new test case to SelectorHandlerUtilTest that specifically reproduces the malformed HQL issue described in the PR title (where @additional_filters@ is preceded by 'and').

```java
 * Covers: processValue, convertRowToJSONObject, fulfillSessionsVariables,
 * convertToHashMap, getNormalizedFieldName, getTargetKey.
 * Includes tests for HQL @additional_filters@ malformation fixes.
```

**Rationale**: The PR title indicates a fix for malformed HQL involving @additional_filters@, but no unit test for this specific scenario was added to the test suite. Adding a regression test ensures the fix works as expected and prevents future regressions.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

**Review by etendobot — APPROVED** [review]
